### PR TITLE
miniupnpc: 2.2.7 -> 2.2.8, again

### PR DESCRIPTION
**Pull request by emilazy**
## Description of changes

Like #325273, except I migrated like half of the reverse dependencies in the universe with my own bare hands and added upstream patches and updates for the other half. My own patches have been sent off to every upstream that still has even the slightest hint of still being maintained.

Every package that built before still builds now, except for `masari`, which I dropped with the consent of its maintainer due to being a dead cryptocurrency abandoned by its upstream. Not that there aren’t a lot of packages that did get patched here that aren’t in much better shape.

There were a lot of real staring‐into‐the‐abyss moments in this one. Hell is other people’s build systems. I’m going to bed now.

cc @nagy @K900 @doronbehar

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by K900** [comment]
You're insane. I just want you to know that.

File: pkgs/applications/networking/p2p/retroshare/default.nix
Comment (doronbehar) [inline pkgs/applications/networking/p2p/retroshare/default.nix:1]: Why is `fetchpatch2` used here?

File: pkgs/applications/blockchains/monero-cli/default.nix
Comment (doronbehar) [inline pkgs/applications/blockchains/monero-cli/default.nix:60]: A bit weird to me how these external patches don't allow us to not have also a local patch... However this shouldn't block this PR.

File: pkgs/by-name/ps/pshs/package.nix
Comment (doronbehar) [inline pkgs/by-name/ps/pshs/package.nix:45]: Is this really needed? Updating this package should be trivial for both `nix{,pkgs}-update`.

**Review by doronbehar — APPROVED** [review]
Looks great to me! I trust you have tested every package touched by this.

**Comment by K900** [comment]
Result of `nixpkgs-review pr 326402` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages marked as broken and skipped:</summary>
  <ul>
    <li>aeon</li>
    <li>bitcoin-unlimited</li>
    <li>bitcoind-unlimited</li>
    <li>groestlcoin</li>
    <li>pivx</li>
    <li>pivxd</li>
  </ul>
</details>
<details>
  <summary>66 packages built:</summary>
  <ul>
    <li>aerc</li>
    <li>alephone</li>
    <li>alephone.icons</li>
    <li>atomic-swap</li>
    <li>bitcoin</li>
    <li>bitcoin-abc</li>
    <li>bitcoind</li>
    <li>bitcoind-abc</li>
    <li>bitcoind-knots</li>
    <li>chiaki4deck</li>
    <li>dante</li>
    <li>dolphin-emu</li>
    <li>dolphin-emu-primehack</li>
    <li>drawpile</li>
    <li>eiskaltdcpp</li>
    <li>elements</li>
    <li>elementsd</li>
    <li>elementsd-simplicity</li>
    <li>flycast</li>
    <li>fragments</li>
    <li>gridcoin-research</li>
    <li>groestlcoind</li>
    <li>haven-cli</li>
    <li>haven-cli.source</li>
    <li>hydrus</li>
    <li>hydrus.doc</li>
    <li>i2pd</li>
    <li>libtransmission_3</li>
    <li>libtransmission_3.apparmor</li>
    <li>libtransmission_4</li>
    <li>libtransmission_4.apparmor</li>
    <li>litecoin</li>
    <li>litecoind</li>
    <li>lutris</li>
    <li>lutris-free</li>
    <li>miniupnpc</li>
    <li>monero-cli</li>
    <li>monero-cli.source</li>
    <li>monero-gui</li>
    <li>namecoind</li>
    <li>particl-core</li>
    <li>pshs</li>
    <li>qodem</li>
    <li>retroshare</li>
    <li>sunshine</li>
    <li>swiften</li>
    <li>torrential</li>
    <li>transmission_3</li>
    <li>transmission_3-gtk</li>
    <li>transmission_3-gtk.apparmor</li>
    <li>transmission_3-qt</li>
    <li>transmission_3-qt.apparmor</li>
    <li>transmission_3.apparmor</li>
    <li>transmission_3_noSystemd</li>
    <li>transmission_3_noSystemd.apparmor</li>
    <li>transmission_4</li>
    <li>transmission_4-gtk</li>
    <li>transmission_4-gtk.apparmor</li>
    <li>transmission_4-qt (transmission_4-qt5)</li>
    <li>transmission_4-qt.apparmor (transmission_4-qt5.apparmor)</li>
    <li>transmission_4-qt6</li>
    <li>transmission_4-qt6.apparmor</li>
    <li>transmission_4.apparmor</li>
    <li>warzone2100</li>
    <li>yaup</li>
    <li>zeroadPackages.zeroad-unwrapped</li>
  </ul>
</details>

**Review by nagy — COMMENTED** [review]
Thank you, this is impressive work.

File: pkgs/applications/networking/p2p/retroshare/default.nix
Comment (emilazy) [reply pkgs/applications/networking/p2p/retroshare/default.nix:1]: `fetchpatch` is “deprecated”; `fetchpatch2` is meant to be better but is backwards‐incompatible, so you have to opt in.

Ironically we probably need a `fetchpatch3`: https://github.com/NixOS/nixpkgs/issues/266556

File: pkgs/applications/blockchains/monero-cli/default.nix
Comment (emilazy) [reply pkgs/applications/blockchains/monero-cli/default.nix:60]: The previous anti‐vendoring patch was actually not fully functional in terms of specifying the header locations. The upstream PRs that fix miniupnpc 2.2.8 also forced the vendoring even harder. So since it’ll need dealing with anyway on the next version, I just fixed the anti‐vendoring stuff and reduced its size in the process.

File: pkgs/by-name/ps/pshs/package.nix
Comment (emilazy) [reply pkgs/by-name/ps/pshs/package.nix:45]: Honestly, it’s never been clear to me exactly where `nixpkgs-update` will update from. I know the main source is Repology, but that it can maybe also do GitHub releases? In this case though the newer releases are just tags and there are no corresponding GitHub releases. This at least lets `maintainers/script/update.nix` work where it doesn’t otherwise.

I asked about best practices [on Matrix](https://matrix.to/#/!kjdutkOsheZdjqYmqp:nixos.org/$UZm4T96zi4OSr7IUL9sq9Rot3olVYz7j50o33R2bU5E?via=lossy.network&via=matrix.org&via=nixos.dev) recently but there wasn’t really any answer. There has been some talk about either making `nix-update-script` the default, or doing a mass addition of it; see https://github.com/NixOS/nixpkgs/pull/325074.

**Comment by emilazy** [comment]
> Looks great to me! I trust you have tested every package touched by this.

Uhh, I tested that they build at least… 🫠

The breaking API change is just two additional parameters that can be set to `NULL, 0` and a renumbering of the return codes, so as long as it builds and any applicable return code checking is updated it should just work. I didn’t try and exercise the functionality of every single package here because there’s way too many of them and “stuff builds” tends to be our rubric for doing updates of libraries anyway. Thankfully a lot of packages had already patched this upstream, so if anything breaks from incorporating those patches it’s on them :)

Of course, I’m happy to be held responsible if any regressions do occur and help out with fixing them, and if there’s any substantive upstream feedback on the patches I wrote I’ll make sure to address them in Nixpkgs as well.

**Comment by miniupnp** [comment]
@emilazy sorry for the API break :(

**Comment by emilazy** [comment]
Hey, it happens :) Although a larger version bump would probably make it catch fewer people by surprise next time.

**Pull request merged**
